public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx

```diff
@@ -0,0 +1,526 @@
+
+  onClickAccept = () => {
+    const selector = buildSelector(this.state.labels);
+    this.props.onChange(selector);
+  };
+
+  onClickClear = () => {
+    this.setState(state => {
+      const labels: SelectableLabel[] = state.labels.map(label => ({
+        ...label,
+        values: undefined,
+        selected: false,
+        loading: false,
+        hidden: false,
+        facets: undefined,
+      }));
+      return { labels, searchTerm: '' };
+    });
+    store.delete(LAST_USED_LABELS_KEY);
+  };
+
+  onClickLabel = (name: string, value: string | undefined, event: React.MouseEvent<HTMLElement>) => {
+    const label = this.state.labels.find(l => l.name === name);
+    if (!label) {
+      return;
+    }
+    // Toggle selected state
+    const selected = !label.selected;
+    let nextValue: Partial<SelectableLabel> = { selected };
+    if (label.values && !selected) {
+      // Deselect all values if label was deselected
+      const values = label.values.map(value => ({ ...value, selected: false }));
+      nextValue = { ...nextValue, facets: 0, values };
+    }
+    this.updateLabelState(name, nextValue, () => {
+      const selectedLabels = this.state.labels.filter(label => label.selected).map(label => label.name);
+      store.setObject(LAST_USED_LABELS_KEY, selectedLabels);
+      if (selected) {
+        // Refetch values for newly selected label...
+        if (!label.values) {
+          this.fetchValues(name);
+        }
+      } else {
+        // Only need to facet when deselecting labels
+        this.doFacetting();
+      }
+    });
+  };
+
+  onClickValue = (name: string, value: string | undefined, event: React.MouseEvent<HTMLElement>) => {
+    const label = this.state.labels.find(l => l.name === name);
+    if (!label || !label.values) {
+      return;
+    }
+    // Toggling value for selected label, leaving other values intact
+    const values = label.values.map(v => ({ ...v, selected: v.name === value ? !v.selected : v.selected }));
+    this.updateLabelState(name, { values }, () => this.doFacetting(name));
+  };
+
+  updateLabelState(name: string, updatedFields: Partial<SelectableLabel>, cb?: () => void) {
+    this.setState(state => {
+      const labels: SelectableLabel[] = state.labels.map(label => {
+        if (label.name === name) {
+          return { ...label, ...updatedFields };
+        }
+        return label;
+      });
+      return { labels };
+    }, cb);
+  }
+
+  componentDidMount() {
+    const { languageProvider } = this.props;
+    if (languageProvider) {
+      const selectedLabels: string[] = store.getObject(LAST_USED_LABELS_KEY, []);
+      languageProvider.start().then(() => {
+        const labels: SelectableLabel[] = languageProvider
+          .getLabelKeys()
+          .slice(0, MAX_LABEL_COUNT)
+          .map(label => ({ name: label, selected: selectedLabels.includes(label), loading: false }));
+        this.setState({ labels }, () => {
+          this.state.labels.forEach(label => {
```

Comment:
I see this on multiple places, couldn't we just fire the `this.fetchValues` on `componentDidUpdate`? Comparing the previous `labels` with current ones and fire the load depending on that? I assume we want to do that every time labels or their selected state changes.